### ISSUE TITLE
HAProxy keeps reloading

### ISSUE THREAD

**Issue by ajibigad** [jcmoraisjr/haproxy-ingress#493]
Hi,

Please, my installation of HAProxy keeps reloading like almost every second.

How can I see the reason the reloads keep happening? Maybe the specific event or configuration change that triggers the reload. I have checked the logs and could not find this.

Here are the logs. It shows that haproxy has been reloaded but does not log what event or configuration change triggered the reload.

```
I0108 05:30:23.975458       8 controller.go:325] Finish HAProxy update id=450: ingress=0.174067ms tcpServices=0.095273ms writeTmpl=1.546662ms reload=24.779582ms total=26.595584ms
I0108 05:31:03.212861       8 controller.go:274] Starting HAProxy update id=451
I0108 05:31:03.238785       8 instance.go:173] HAProxy successfully reloaded
I0108 05:31:03.238852       8 controller.go:325] Finish HAProxy update id=451: ingress=0.177107ms tcpServices=0.077803ms writeTmpl=1.443728ms reload=24.171472ms total=25.870110ms
I0108 05:31:04.904444       8 controller.go:274] Starting HAProxy update id=452
I0108 05:31:04.934144       8 instance.go:173] HAProxy successfully reloaded
I0108 05:31:04.934217       8 controller.go:325] Finish HAProxy update id=452: ingress=0.255840ms tcpServices=0.085752ms writeTmpl=1.358251ms reload=27.943704ms total=29.643547ms
I0108 05:31:06.904449       8 controller.go:274] Starting HAProxy update id=453
I0108 05:31:06.931716       8 instance.go:173] HAProxy successfully reloaded
I0108 05:31:06.931769       8 controller.go:325] Finish HAProxy update id=453: ingress=0.302308ms tcpServices=0.082244ms writeTmpl=0.979577ms reload=25.846153ms total=27.210282ms
I0108 05:31:13.823594       8 controller.go:274] Starting HAProxy update id=454
I0108 05:31:13.848823       8 instance.go:173] HAProxy successfully reloaded
I0108 05:31:13.848878       8 controller.go:325] Finish HAProxy update id=454: ingress=0.170323ms tcpServices=0.079858ms writeTmpl=1.166887ms reload=23.748202ms total=25.165270ms
I0108 05:31:14.904386       8 controller.go:274] Starting HAProxy update id=455
I0108 05:31:14.931440       8 instance.go:173] HAProxy successfully reloaded
I0108 05:31:14.931545       8 controller.go:325] Finish HAProxy update id=455: ingress=0.216895ms tcpServices=0.126053ms writeTmpl=1.571408ms reload=25.087125ms total=27.001481ms
I0108 05:31:16.904418       8 controller.go:274] Starting HAProxy update id=456
I0108 05:31:16.905052       8 instance.go:164] old and new configurations match
I0108 05:31:16.905095       8 controller.go:325] Finish HAProxy update id=456: ingress=0.233936ms tcpServices=0.083336ms total=0.317272ms
I0108 05:31:48.712694       8 controller.go:274] Starting HAProxy update id=457
I0108 05:31:48.713796       8 instance.go:164] old and new configurations match
I0108 05:31:48.713862       8 controller.go:325] Finish HAProxy update id=457: ingress=0.192103ms tcpServices=0.262257ms total=0.454360ms
I0108 05:31:51.927229       8 controller.go:274] Starting HAProxy update id=458
I0108 05:31:51.927800       8 instance.go:164] old and new configurations match
I0108 05:31:51.927835       8 controller.go:325] Finish HAProxy update id=458: ingress=0.169422ms tcpServices=0.128642ms total=0.298064ms
I0108 05:31:56.110371       8 controller.go:274] Starting HAProxy update id=459
I0108 05:31:56.111132       8 instance.go:164] old and new configurations match
I0108 05:31:56.112250       8 controller.go:325] Finish HAProxy update id=459: ingress=0.385960ms tcpServices=0.095490ms total=0.481450ms
I0108 05:32:10.955647       8 controller.go:274] Starting HAProxy update id=460
I0108 05:32:10.986820       8 instance.go:173] HAProxy successfully reloaded
I0108 05:32:10.986881       8 controller.go:325] Finish HAProxy update id=460: ingress=0.210989ms tcpServices=0.090432ms writeTmpl=1.553318ms reload=29.269376ms total=31.124115ms
I0108 05:32:38.693664       8 controller.go:274] Starting HAProxy update id=461
I0108 05:32:38.722752       8 instance.go:173] HAProxy successfully reloaded
I0108 05:32:38.722858       8 controller.go:325] Finish HAProxy update id=461: ingress=0.170117ms tcpServices=0.086624ms writeTmpl=1.724356ms reload=27.040588ms total=29.021685ms
I0108 05:32:50.878999       8 controller.go:274] Starting HAProxy update id=462
I0108 05:32:50.879620       8 instance.go:164] old and new configurations match
I0108 05:32:50.879700       8 controller.go:325] Finish HAProxy update id=462: ingress=0.170757ms tcpServices=0.083201ms total=0.253958ms
I0108 05:33:45.652694       8 controller.go:274] Starting HAProxy update id=463
I0108 05:33:45.685435       8 instance.go:173] HAProxy successfully reloaded
I0108 05:33:45.685615       8 controller.go:325] Finish HAProxy update id=463: ingress=0.169460ms tcpServices=0.083672ms writeTmpl=1.449638ms reload=30.971717ms total=32.674487ms
I0108 05:35:30.734393       8 controller.go:274] Starting HAProxy update id=464
I0108 05:35:30.765739       8 instance.go:173] HAProxy successfully reloaded
I0108 05:35:30.765804       8 controller.go:325] Finish HAProxy update id=464: ingress=0.267009ms tcpServices=0.115642ms writeTmpl=2.319440ms reload=28.566865ms total=31.268956ms
I0108 05:35:34.935195       8 controller.go:274] Starting HAProxy update id=465
I0108 05:35:34.935943       8 instance.go:164] old and new configurations match
I0108 05:35:34.935983       8 controller.go:325] Finish HAProxy update id=465: ingress=0.283440ms tcpServices=0.090909ms total=0.374349ms
I0108 05:36:17.198922       8 controller.go:274] Starting HAProxy update id=466
I0108 05:36:17.233783       8 instance.go:173] HAProxy successfully reloaded
I0108 05:36:17.233856       8 controller.go:325] Finish HAProxy update id=466: ingress=0.176172ms tcpServices=0.092310ms writeTmpl=1.579722ms reload=32.921219ms total=34.769423ms
I0108 05:36:33.890393       8 controller.go:274] Starting HAProxy update id=467
I0108 05:36:33.916625       8 instance.go:173] HAProxy successfully reloaded
I0108 05:36:33.916681       8 controller.go:325] Finish HAProxy update id=467: ingress=0.220096ms tcpServices=0.099175ms writeTmpl=1.500365ms reload=24.361458ms total=26.181094ms
I0108 05:36:34.904405       8 controller.go:274] Starting HAProxy update id=468
I0108 05:36:34.932050       8 instance.go:173] HAProxy successfully reloaded
I0108 05:36:34.932145       8 controller.go:325] Finish HAProxy update id=468: ingress=0.216520ms tcpServices=0.098154ms writeTmpl=1.427979ms reload=25.849831ms total=27.592484ms
I0108 05:36:36.904457       8 controller.go:274] Starting HAProxy update id=469
I0108 05:36:36.933307       8 instance.go:173] HAProxy successfully reloaded
I0108 05:36:36.933359       8 controller.go:325] Finish HAProxy update id=469: ingress=0.270744ms tcpServices=0.089969ms writeTmpl=1.461116ms reload=26.929931ms total=28.751760ms
I0108 05:36:43.823283       8 controller.go:274] Starting HAProxy update id=470
I0108 05:36:43.823882       8 instance.go:164] old and new configurations match
I0108 05:36:43.824084       8 controller.go:325] Finish HAProxy update id=470: ingress=0.238361ms tcpServices=0.086363ms total=0.324724ms
I0108 05:36:44.904434       8 controller.go:274] Starting HAProxy update id=471
I0108 05:36:44.905062       8 instance.go:164] old and new configurations match
I0108 05:36:44.905128       8 controller.go:325] Finish HAProxy update id=471: ingress=0.265472ms tcpServices=0.088157ms total=0.353629ms
I0108 05:36:46.904436       8 controller.go:274] Starting HAProxy update id=472
I0108 05:36:46.905049       8 instance.go:164] old and new configurations match
I0108 05:36:46.905090       8 controller.go:325] Finish HAProxy update id=472: ingress=0.247689ms tcpServices=0.090032ms total=0.337721ms
I0108 05:37:16.735031       8 controller.go:274] Starting HAProxy update id=473
I0108 05:37:16.767818       8 instance.go:173] HAProxy successfully reloaded
I0108 05:37:16.767878       8 controller.go:325] Finish HAProxy update id=473: ingress=0.180940ms tcpServices=0.089903ms writeTmpl=3.876402ms reload=28.590624ms total=32.737869ms
I0108 05:37:38.990807       8 controller.go:274] Starting HAProxy update id=474
I0108 05:37:39.021337       8 instance.go:173] HAProxy successfully reloaded
I0108 05:37:39.021431       8 controller.go:325] Finish HAProxy update id=474: ingress=0.234313ms tcpServices=0.091422ms writeTmpl=1.550026ms reload=28.595411ms total=30.471172ms
I0108 05:37:40.904419       8 controller.go:274] Starting HAProxy update id=475
I0108 05:37:40.933743       8 instance.go:173] HAProxy successfully reloaded
I0108 05:37:40.933800       8 controller.go:325] Finish HAProxy update id=475: ingress=0.208523ms tcpServices=0.088654ms writeTmpl=1.490073ms reload=27.474624ms total=29.261874ms
I0108 05:37:51.112371       8 controller.go:274] Starting HAProxy update id=476
I0108 05:37:51.140506       8 instance.go:173] HAProxy successfully reloaded
I0108 05:37:51.140787       8 controller.go:325] Finish HAProxy update id=476: ingress=0.218311ms tcpServices=0.122190ms writeTmpl=1.354439ms reload=26.368082ms total=28.063022ms
I0108 05:38:21.514935       8 controller.go:274] Starting HAProxy update id=477
I0108 05:38:21.548299       8 instance.go:173] HAProxy successfully reloaded
I0108 05:38:21.548486       8 controller.go:325] Finish HAProxy update id=477: ingress=0.195362ms tcpServices=0.096912ms writeTmpl=1.586844ms reload=31.430324ms total=33.309442ms
I0108 05:38:41.793872       8 controller.go:274] Starting HAProxy update id=478
I0108 05:38:41.794458       8 instance.go:164] old and new configurations match
I0108 05:38:41.794558       8 controller.go:325] Finish HAProxy update id=478: ingress=0.224326ms tcpServices=0.095540ms total=0.319866ms
```

Thanks

**Comment by jcmoraisjr** [comment]
Hi, you can follow one (or two!) of these paths:

* Add `--v=2` command-line option to increase the log verbosity. Its verbosity is very low, btw I'm considering leave it as the default. You should see the log in the end of [this if statement](https://github.com/jcmoraisjr/haproxy-ingress/blob/v0.8/pkg/haproxy/dynupdate.go#L86);
* Exec `kubectl get xxx --all-namespaces -w` where `xxx` is one of the k8s resources you can find [here](https://github.com/jcmoraisjr/haproxy-ingress/blob/v0.8/pkg/common/ingress/controller/listers.go#L224) -- Ingress, Endpoint, Secret, ConfigMap, Service and Pod. One of them should be very noisy.

Please share your findinds, perhaps we can improve something in the lister or in the dynamic update.

Hth.

**Comment by ajibigad** [comment]
Hi @jcmoraisjr 

Thanks for replying.
These are my findings.

I added this command-line-arg `- --watch-namespace=mynamespace` to make sure only resources in this namespace are watched.

I watched all these resources (Ingress, Endpoint, Secret, ConfigMap, Service, Pod and Node) and the noisy resources are configmaps caused by ingress-controller-leader-haproxy configmap and nodes.
`kubectl get configmaps -w -n mynamespace`
```
<After logging other configmaps>
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
ingress-controller-leader-haproxy             0      31h
```
I also noticed nodes might be watched too as written [here](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/common/ingress/controller/listers.go#L252). My nodes watch logs are also noisy.

I don't mind the noise, the issue is it reloads the haproxy even when i cannot see any difference between the old and new config that is worth reloading haproxy

After adding the command line option `--v=2` these are the logs.
```
I0108 19:31:56.533173       8 status.go:193] I am the new status update leader
I0108 19:33:26.519425       8 controller.go:274] Starting HAProxy update id=4
I0108 19:33:26.520205       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:33:26.552226       8 instance.go:173] HAProxy successfully reloaded
I0108 19:33:26.552291       8 controller.go:325] Finish HAProxy update id=4: ingress=0.551753ms tcpServices=0.091674ms writeTmpl=2.660975ms reload=29.436833ms total=32.741235ms
I0108 19:33:39.543487       8 controller.go:274] Starting HAProxy update id=5
I0108 19:33:39.543958       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:33:39.581437       8 instance.go:173] HAProxy successfully reloaded
I0108 19:33:39.581707       8 controller.go:325] Finish HAProxy update id=5: ingress=0.190154ms tcpServices=0.092374ms writeTmpl=4.376015ms reload=33.190710ms total=37.849253ms
I0108 19:33:40.742694       8 controller.go:274] Starting HAProxy update id=6
I0108 19:33:40.743238       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:33:40.783443       8 instance.go:173] HAProxy successfully reloaded
I0108 19:33:40.783537       8 controller.go:325] Finish HAProxy update id=6: ingress=0.237386ms tcpServices=0.130216ms writeTmpl=1.687510ms reload=38.639778ms total=40.694890ms
I0108 19:33:48.763410       8 controller.go:274] Starting HAProxy update id=7
I0108 19:33:48.763808       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:33:48.788797       8 instance.go:173] HAProxy successfully reloaded
I0108 19:33:48.788874       8 controller.go:325] Finish HAProxy update id=7: ingress=0.161646ms tcpServices=0.092763ms writeTmpl=1.475613ms reload=23.611409ms total=25.341431ms
I0108 19:36:15.920462       8 controller.go:274] Starting HAProxy update id=8
I0108 19:36:15.921198       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:36:15.955162       8 instance.go:173] HAProxy successfully reloaded
I0108 19:36:15.955217       8 controller.go:325] Finish HAProxy update id=8: ingress=0.460370ms tcpServices=0.109582ms writeTmpl=1.582537ms reload=32.469345ms total=34.621834ms
I0108 19:36:20.976312       8 controller.go:274] Starting HAProxy update id=9
I0108 19:36:20.976711       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:36:21.007735       8 instance.go:173] HAProxy successfully reloaded
I0108 19:36:21.007903       8 controller.go:325] Finish HAProxy update id=9: ingress=0.168352ms tcpServices=0.103565ms writeTmpl=1.813602ms reload=29.288928ms total=31.374447ms
I0108 19:36:54.462293       8 controller.go:274] Starting HAProxy update id=10
I0108 19:36:54.462988       8 instance.go:164] old and new configurations match
I0108 19:36:54.463037       8 controller.go:325] Finish HAProxy update id=10: ingress=0.207457ms tcpServices=0.119773ms total=0.327230ms
I0108 19:37:18.741209       8 controller.go:274] Starting HAProxy update id=11
I0108 19:37:18.741782       8 instance.go:164] old and new configurations match
I0108 19:37:18.741824       8 controller.go:325] Finish HAProxy update id=11: ingress=0.154386ms tcpServices=0.080726ms total=0.235112ms
I0108 19:37:19.757766       8 controller.go:274] Starting HAProxy update id=12
I0108 19:37:19.758212       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:37:19.791211       8 instance.go:173] HAProxy successfully reloaded
I0108 19:37:19.791302       8 controller.go:325] Finish HAProxy update id=12: ingress=0.187631ms tcpServices=0.086394ms writeTmpl=1.658441ms reload=31.466460ms total=33.398926ms
I0108 19:37:40.923590       8 controller.go:274] Starting HAProxy update id=13
I0108 19:37:40.923965       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:37:40.963621       8 instance.go:173] HAProxy successfully reloaded
I0108 19:37:40.963692       8 controller.go:325] Finish HAProxy update id=13: ingress=0.157521ms tcpServices=0.082066ms writeTmpl=2.514406ms reload=37.224017ms total=39.978010ms
I0108 19:38:56.624223       8 controller.go:274] Starting HAProxy update id=14
I0108 19:38:56.624794       8 dynupdate.go:100] diff outside backends - [tcp-services]
I0108 19:38:56.652820       8 instance.go:173] HAProxy successfully reloaded
I0108 19:38:56.652875       8 controller.go:325] Finish HAProxy update id=14: ingress=0.261258ms tcpServices=0.098423ms writeTmpl=1.407934ms reload=26.707208ms total=28.474823ms
```

I can see the diff is `dynupdate.go:100] diff outside backends - [tcp-services]`

It's strange to me that the TCP backends differ.

I ran a shell in the haproxy-ingress pod to monitor the `haproxy.cfg` file to see the changes after each reload. I copied the contents of the config file to another file `haproxy-temp.cfg` then ran a `diff` on both files after a reload. The only change I noticed was the arrangement of the TCP backend services and I believed this is cause this flag is false by default `--sort-backends`. 
I added the `- --sort-backends=true` to the haproxy-ingress deployment yaml but the TCP backend services are still randomized and not sorted.

I think the TCP backends should only differ if the backend services or their endpoints changes or a new backend service is added and not their order. What do you think? I think this is the problem.

**Comment by jcmoraisjr** [comment]
HI, in fact the tcp services endpoints should be the problem here - it isn't neither sorted or analyzed in the dynamic config module. Can you confirm if you see reloads whenever the tcp service endpoints changes its order and the reload doesn't happen when the order doesn't change? I'll have a proper look on this, thanks for reporting.

In the mean time I think you can workaround this by applying this patch:

```diff
diff --git a/pkg/converters/configmap/tcpservices.go b/pkg/converters/configmap/tcpservices.go
index b80772a..86e664e 100644
--- a/pkg/converters/configmap/tcpservices.go
+++ b/pkg/converters/configmap/tcpservices.go
@@ -18,6 +18,7 @@ package configmap

 import (
 	"fmt"
+	"sort"
 	"strconv"
 	"strings"

@@ -92,6 +93,9 @@ func (c *tcpSvcConverter) Sync(tcpservices map[string]string) {
 		}
 		servicename := fmt.Sprintf("%s_%s", service.Namespace, service.Name)
 		backend := c.haproxy.AcquireTCPBackend(servicename, publicport)
+		sort.Slice(addrs, func(i, j int) bool {
+			return addrs[i].IP < addrs[j].IP
+		})
 		for _, addr := range addrs {
 			backend.AddEndpoint(addr.IP, addr.Port)
 		}
```

**Comment by ajibigad** [comment]
Hi @jcmoraisjr 

I think the only way I can confirm the differences is by logging the old and current config tcpBackends before comparing them [here.](https://github.com/jcmoraisjr/haproxy-ingress/blob/1c286207f6319a78904bc48bca7f1025b75faf28/pkg/haproxy/dynupdate.go#L94)

Please, can you release a patch that logs these?

Also, what does the patch you shared do? It looks like it would sort the endpoints for each TCP backend service. 

This is the result of the diff I did on the config files in the pod.
```
  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #	  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #	# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #								# #
# #   TCP SERVICES						# #   TCP SERVICES
# #								# #
#								#
listen _tcp_mynamespace_custom-service_7900		      <
    bind *:7900 ssl crt /ingress-controller/ssl/mynamespace_fr <
    mode tcp						      <
    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ % <
    server srv001 10.20.0.82:7900			      <
    server srv002 10.20.2.29:7900			      <
    server srv003 10.20.5.58:7900			      <
    server srv004 10.20.7.10:7900			      <
listen _tcp_mynamespace_custom-service_7901			listen _tcp_mynamespace_custom-service_7901
    bind *:7901 ssl crt /ingress-controller/ssl/mynamespace_fr	    bind *:7901 ssl crt /ingress-controller/ssl/mynamespace_fr
    mode tcp							    mode tcp
    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %	    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %
    server srv001 10.20.0.82:7901				    server srv001 10.20.0.82:7901
    server srv002 10.20.2.29:7901				    server srv002 10.20.2.29:7901
    server srv003 10.20.5.58:7901				    server srv003 10.20.5.58:7901
    server srv004 10.20.7.10:7901				    server srv004 10.20.7.10:7901
listen _tcp_mynamespace_custom-service_7902			listen _tcp_mynamespace_custom-service_7902
    bind *:7902 ssl crt /ingress-controller/ssl/mynamespace_fr	    bind *:7902 ssl crt /ingress-controller/ssl/mynamespace_fr
    mode tcp							    mode tcp
    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %	    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %
    server srv001 10.20.0.82:7902				    server srv001 10.20.0.82:7902
    server srv002 10.20.2.29:7902				    server srv002 10.20.2.29:7902
    server srv003 10.20.5.58:7902				    server srv003 10.20.5.58:7902
    server srv004 10.20.7.10:7902				    server srv004 10.20.7.10:7902
listen _tcp_mynamespace_custom-service_7903			listen _tcp_mynamespace_custom-service_7903
    bind *:7903 ssl crt /ingress-controller/ssl/mynamespace_fr	    bind *:7903 ssl crt /ingress-controller/ssl/mynamespace_fr
    mode tcp							    mode tcp
    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %	    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %
    server srv001 10.20.0.82:7903				    server srv001 10.20.0.82:7903
    server srv002 10.20.2.29:7903				    server srv002 10.20.2.29:7903
    server srv003 10.20.5.58:7903				    server srv003 10.20.5.58:7903
    server srv004 10.20.7.10:7903				    server srv004 10.20.7.10:7903
							      >	listen _tcp_mynamespace_custom-service_7900
							      >	    bind *:7900 ssl crt /ingress-controller/ssl/mynamespace_fr
							      >	    mode tcp
							      >	    log-format %ci:%cp\ [%t]\ %ft\ %b/%s\ %Tw/%Tc/%Tt\ %Th\ %
							      >	    server srv001 10.20.0.82:7900
							      >	    server srv002 10.20.2.29:7900
							      >	    server srv003 10.20.5.58:7900
							      >	    server srv004 10.20.7.10:7900
```
If you notice, the difference is the order of the TCP backend services. The endpoints per backend service are still in the same order as they were before the reload.

**Comment by jcmoraisjr** [comment]
> This is the result of the diff I did on the config files in the pod.

Hi, thanks for this output! Names isn't being properly sorted as well, this another patch should solve this:

```diff
diff --git a/pkg/haproxy/config.go b/pkg/haproxy/config.go
index 99779dc..203e093 100644
--- a/pkg/haproxy/config.go
+++ b/pkg/haproxy/config.go
@@ -102,7 +102,12 @@ func (c *config) AcquireTCPBackend(servicename string, port int) *hatypes.TCPBac
 	}
 	c.tcpbackends = append(c.tcpbackends, backend)
 	sort.Slice(c.tcpbackends, func(i, j int) bool {
-		return c.tcpbackends[i].Name < c.tcpbackends[j].Name
+		back1 := c.tcpbackends[i]
+		back2 := c.tcpbackends[j]
+		if back1.Name == back2.Name {
+			return back1.Port < back2.Port
+		}
+		return back1.Name < back2.Name
 	})
 	return backend
 }
```

> I added the - --sort-backends=true to the haproxy-ingress deployment yaml but the TCP backend services are still randomized and not sorted.

Yep, I can confirm sort-backends wasn't reimplemented in the new controller, thanks for pointing this.

> Please, can you release a patch that logs these?

I think this isn't needed anymore, your diff output is clear enough.

> Also, what does the patch you shared do? It looks like it would sort the endpoints for each TCP backend service.

Yes, the first patch sorts the endpoints but its just a workaround. The right fix must be in the dynconfig.

**Comment by ajibigad** [comment]
Hi @jcmoraisjr 

Thanks for the support so far. Really appreciate.

After some studying of the codebase, I think I understand the source of the problem now.

During an update, the [SyncIngress function](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/controller/controller.go#L269) is called.
It then calls the `GetByName` function at this [line](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/controller/controller.go#L307) to get the tcpconfigmap.
This returns a k8s object type `ConfigMap` 
```
// ConfigMap holds configuration data for components or applications to consume.
type ConfigMap struct {
	metav1.TypeMeta
	// +optional
	metav1.ObjectMeta

	// Data contains the configuration data.
	// Each key must consist of alphanumeric characters, '-', '_' or '.'.
	// +optional
	Data map[string]string
}
```
The configmap's `Data` field is then passed to the [tcpSvcConverter's sync](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/converters/configmap/tcpservices.go#L50). As you can see in the struct, field `Data` is a `map`.
The rearrangement of the tcpbackendservices is due to two reasons

1. The nature of maps in golang.
> When iterating over a map with a range loop, the iteration order is not specified and is not guaranteed to be the same from one iteration to the next. [Source: Golang Doc](https://blog.golang.org/go-maps-in-action) 

2. In my case, the service name is the same for all my TCP backend services, only the ports differ.
This [line](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/haproxy/config.go#L96) attempts to sort the tcpbackendservices using only the names of the services and not the ports.

The endpoints would have been an issue too cause they are not sorted after adding them to the backend struct but so far they come in the same order at this [line](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/converters/configmap/tcpservices.go#L80).

I wrote a small test [here](https://play.golang.org/p/9gDE93gTcL8) to highlight the result of `reflect.DeepEqual` when the order of the TCP backend services or their endpoints changes in the config. It returns `false` in these cases. I believe this would be the same result [here](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/haproxy/dynupdate.go#L91).

Adding both patches you suggested should solve the problem. One to sort the backend services using their ports if the names are the same. The other to sort the endpoints by IP just in case.

I would have loved to create a pull request with these changes and tests for them but I am not really familiar with golang and it might take some time to get a grasp of the language. Please, can you release the patch as a new 0.8 tag on your docker repository so myself and others that might face this same challenge can use that tag in our deployment?

Please let me know if this would be possible. Thanks.

Also, I noticed the [doc for 0.8](https://github.com/jcmoraisjr/haproxy-ingress/tree/release-0.8#reload-strategy) says the default reload strategy is `native` but from the code [here](https://github.com/jcmoraisjr/haproxy-ingress/blob/1351a73e8f7732ab9765aa90619b5246cc456c55/pkg/controller/controller.go#L209), it is `reusesocket`. Please, can you confirm this?

Thanks again.

**Comment by jcmoraisjr** [comment]
Hi, your description perfectly describes what's happening. This will be fixed for sure, btw I'm relabeling this issue to better reflect that this isn't just an improvement. The fix will also be backported to v0.8.

The backend sorting is just that and it's quite simple. The endpoint one should also be fixed in the dynconfig and this isn't that simple. Anyway I'll start just after finish some metrics docs and will tag a new v0.8 in a few days. In the mean time you can clone release-0.8, patch the code and build a new image yourself, it's really simple - just place docker and go 1.9.2 in the path and issue `make build test image`. I'll notify here as soon as the fix is tagged.

**Comment by ajibigad** [comment]
Hi @jcmoraisjr 

I have been able to build and deploy the patch and it worked fine.
Another workaround was to create new kubernetes services for each of the ports to enable sorting by name work correctly.

I believe you would close the issue once the fix is officially tagged.

Thanks again for your help in resolving this.
